src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -0,0 +1,691 @@
+                //         delegate*<object, void> ptr2 = param1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param1").WithArguments("delegate*<ref object,void>", "delegate*<object,void>").WithLocation(7, 40),
+                // (8,44): error CS0266: Cannot implicitly convert type 'delegate*<ref object,void>' to 'delegate*<ref string,void>'. An explicit conversion exists (are you missing a cast?)
+                //         delegate*<ref string, void> ptr3 = param1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param1").WithArguments("delegate*<ref object,void>", "delegate*<ref string,void>").WithLocation(8, 44)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerToFunctionPointerInvalid_ReturnTypeVariance()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(delegate*<ref string> param1)
+    {
+        delegate*<ref readonly string> ptr1 = param1;
+        delegate*<string> ptr2 = param1;
+        delegate*<object> ptr3 = param1;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,47): error CS0266: Cannot implicitly convert type 'delegate*<string>' to 'delegate*<string>'. An explicit conversion exists (are you missing a cast?)
+                //         delegate*<ref readonly string> ptr1 = param1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param1").WithArguments("delegate*<string>", "delegate*<string>").WithLocation(6, 47),
+                // (7,34): error CS0266: Cannot implicitly convert type 'delegate*<string>' to 'delegate*<string>'. An explicit conversion exists (are you missing a cast?)
+                //         delegate*<string> ptr2 = param1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param1").WithArguments("delegate*<string>", "delegate*<string>").WithLocation(7, 34),
+                // (8,34): error CS0266: Cannot implicitly convert type 'delegate*<string>' to 'delegate*<object>'. An explicit conversion exists (are you missing a cast?)
+                //         delegate*<object> ptr3 = param1;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param1").WithArguments("delegate*<string>", "delegate*<object>").WithLocation(8, 34)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerToFunctionPointerInvalid_CallingConvention()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(delegate*<void> param1)
+    {
+        delegate* cdecl<void> ptr1 = param1;
+        delegate* thiscall<void> ptr2 = param1;
+        delegate* stdcall<void> ptr3 = param1;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,38): error CS0266: Cannot implicitly convert type 'delegate*<void>' to 'delegate*<void>'. An explicit conversion exists (are you missing a cast?)
```

Comment:
>'delegate*<void>' to 'delegate*<void>' [](start = 72, length = 38)

The types look the same.

Reply:
There's an existing prototype comment to include the calling convention in error formatting, this message will change when that's fixed.